### ISSUE TITLE
Recursion in kernels should be allowed in a compile-time constexpr evaluation context

### ISSUE THREAD

**Issue by keryell** [KhronosGroup/SYCL-Docs#267]
https://github.com/intel/llvm/issues/6479 mentions that something like
```c++
auto constexpr a = some_recursive_function();
```
is not allowed in DPC++ because it strictly follows the specification SYCL 2020 rev. 5 https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:language.restrictions.kernels

> 5.4. Language restrictions for device functions
> [...]
> - Recursion is not allowed in a device function

but since more and more things can be evaluated at compile time in modern C++ (`constexpr`, then `constinit` and `consteval`), all these restrictions should be waived in the case of a compile-time evaluation because it would break any modern use of metaprogramming.
There is still the quality-of-implementation of some `constexpr` evaluations which are allowed to happen at runtime, I have the feeling that for the quality of life of a user we should enforce a SYCL compliant implementation to do it a compile-time if it can be done.
This is how I interpret the spirit of SYCL in the context of modern C++.

**Comment by Naghasan** [comment]
> all these restrictions should be waived in the case of a compile-time evaluation because it would break any modern use of metaprogramming

+1, these restrictions are there to reflect device limitations, no reason it should impact compile time evaluation.

**Comment by keryell** [comment]
> On a side note, cuda allows recursion. Should we relax the `Recursion is not allowed in a device function`, and transform it a "may", or any people with a cuda backend implementation will implement recursion as a kind of extension?
> 
> Also `gnu` and `llvm` have a attribute to force tail recursion ([clang::musttail](https://clang.llvm.org/docs/AttributeReference.html#id586)), in theory this should allow the code to compiled for the device without probleme

Yes, it seems compelling not to constrain SYCL with an OpenCL mindset if the user is using another backend with more capabilities like those ones.

**Comment by fraggamuffin** [comment]
SHould we remove recursion limitation in general, or just do what the title asks? prefer no limit
also consider const expr function pointer issue during drafting

**Comment by gmlueck** [comment]
I believe this was fixed by #388, so closing this.
